File: qadevOOo/tests/java/ifc/sheet/_SpreadsheetDocumentSettings.java
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */
package ifc.sheet;

import lib.MultiPropertyTest;

import com.sun.star.i18n.ForbiddenCharacters;
import com.sun.star.i18n.XForbiddenCharacters;
import com.sun.star.lang.Locale;
import com.sun.star.uno.UnoRuntime;
import com.sun.star.util.Date;


/**
* Testing <code>com.sun.star.sheet.SpreadsheetDocumentSettings</code>
* service properties :
* <ul>
*  <li><code> IsIterationEnabled</code></li>
*  <li><code> IterationCount</code></li>
*  <li><code> IterationEpsilon</code></li>
*  <li><code> StandardDecimals</code></li>
*  <li><code> NullDate</code></li>
*  <li><code> DefaultTabStop</code></li>
*  <li><code> IgnoreCase</code></li>
*  <li><code> CalcAsShown</code></li>
*  <li><code> MatchWholeCell</code></li>
*  <li><code> SpellOnline</code></li>
*  <li><code> LookUpLabels</code></li>
*  <li><code> RegularExpressions</code></li>
*  <li><code> ForbiddenCharacters</code></li>
*  <li><code> HasDrawPages</code></li>
*  <li><code> Wildcards</code></li>
* </ul> <p>
* Properties testing is automated by <code>lib.MultiPropertyTest</code>.
* @see com.sun.star.sheet.SpreadsheetDocumentSettings
*/
public class _SpreadsheetDocumentSettings extends MultiPropertyTest {
    /**
     *This class is destined to custom test of property <code>NullDate</code>.
     */
    protected PropertyTester DateTester = new PropertyTester() {
        @Override
        protected Object getNewValue(String propName, Object oldValue) {
            Date date = (Date) oldValue;
            Date newDate = new Date((short) (date.Day - 1), date.Month,
                                    date.Year);

            return newDate;
        }
    };

    /**
     *This class is destined to custom test of property <code>ForbiddenCharacters</code>.
     */
    protected PropertyTester ChrTester = new PropertyTester() {
        @Override
        protected Object getNewValue(String propName, Object oldValue) {
            return new ForbiddenChrTest();
        }

        @Override
        protected boolean compare(Object obj1, Object obj2) {
            Locale loc = new Locale("ru", "RU", "");
            XForbiddenCharacters fc1 = UnoRuntime.queryInterface(
                                               XForbiddenCharacters.class,
                                               obj1);
            XForbiddenCharacters fc2 = UnoRuntime.queryInterface(
                                               XForbiddenCharacters.class,
                                               obj2);
            boolean has1 = fc1.hasForbiddenCharacters(loc);
            boolean has2 = fc2.hasForbiddenCharacters(loc);

            return has1 == has2;
        }
    };

    /**
     * Test property <code>NullDate</code> using custom <code>PropertyTest</code>.
     */
    public void _NullDate() {
        testProperty("NullDate", DateTester);
    }

    /**
     * Test property <code>ForbiddenCharacters</code> using custom <code>PropertyTest</code>.
     */
    public void _ForbiddenCharacters() {
        testProperty("ForbiddenCharacters", ChrTester);
    }

    /**
     * Class implements interface <code>XForbiddenCharacters</code>.
     * It's destined to custom test of property <code>ForbiddenCharacters</code>.
     * Feature of the class that it supports forbidden characters
     * for russian locale.
     * @see com.sun.star.i18n.XForbiddenCharacters
     */
    protected static class ForbiddenChrTest implements XForbiddenCharacters {
        protected Locale locale = new Locale("ru", "RU", "");
        protected ForbiddenCharacters chrs = new ForbiddenCharacters("q", "w");

        public ForbiddenCharacters getForbiddenCharacters(Locale rLocale)
            throws com.sun.star.container.NoSuchElementException {
            if (rLocale.Country.equals(locale.Country) &&
                    rLocale.Language.equals(locale.Language) &&
                    rLocale.Variant.equals(locale.Variant)) {
                return chrs;
            }

            throw new com.sun.star.container.NoSuchElementException();
        }

        public void setForbiddenCharacters(Locale rLocale,
                                           ForbiddenCharacters rForbiddenCharacters) {
        }

        public void removeForbiddenCharacters(Locale rLocale) {
        }

        public boolean hasForbiddenCharacters(Locale rLocale) {
            return rLocale.Country.equals(locale.Country) &&
                    rLocale.Language.equals(locale.Language) &&
                    rLocale.Variant.equals(locale.Variant);
        }
    }
}

File: test/source/sheet/spreadsheetdocumentsettings.cxx
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4; fill-column: 100 -*- */
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include <test/unoapi_property_testers.hxx>
#include <test/sheet/spreadsheetdocumentsettings.hxx>

#include <com/sun/star/awt/XDevice.hpp>
#include <com/sun/star/beans/XPropertySet.hpp>
#include <com/sun/star/lang/Locale.hpp>
#include <com/sun/star/i18n/XForbiddenCharacters.hpp>
#include <com/sun/star/util/Date.hpp>
#include <com/sun/star/uno/Any.hxx>
#include <com/sun/star/uno/Reference.hxx>

#include <cppunit/TestAssert.h>

using namespace com::sun::star;
using namespace com::sun::star::uno;

namespace apitest
{
void SpreadsheetDocumentSettings::testSpreadsheetDocumentSettingsProperties()
{
    uno::Reference<beans::XPropertySet> xSpreadsheetDocumentSettings(init(), UNO_QUERY_THROW);
    OUString propName;
    uno::Any aNewValue;

    propName = "IsIterationEnabled";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);

    propName = "IterationCount";
    testLongProperty(xSpreadsheetDocumentSettings, propName);

    propName = "IterationEpsilon";
    testDoubleProperty(xSpreadsheetDocumentSettings, propName);

    propName = "StandardDecimals";
    testShortProperty(xSpreadsheetDocumentSettings, propName);

    propName = "NullDate";
    util::Date aNullDateGet;
    util::Date aNullDateSet;
    CPPUNIT_ASSERT_MESSAGE("Unable to get PropertyValue: NullDate",
                           xSpreadsheetDocumentSettings->getPropertyValue(propName)
                           >>= aNullDateGet);

    aNewValue <<= util::Date(1, 1, 2000);
    xSpreadsheetDocumentSettings->setPropertyValue(propName, aNewValue);
    CPPUNIT_ASSERT(xSpreadsheetDocumentSettings->getPropertyValue(propName) >>= aNullDateSet);
    CPPUNIT_ASSERT_EQUAL_MESSAGE("Unable to set PropertyValue: NullDate", sal_Int16(2000),
                                 aNullDateSet.Year);

    propName = "DefaultTabStop";
    testShortProperty(xSpreadsheetDocumentSettings, propName);

    propName = "IgnoreCase";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);

    propName = "CalcAsShown";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);

    propName = "MatchWholeCell";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);

    propName = "SpellOnline";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);

    propName = "LookUpLabels";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);

    propName = "RegularExpressions";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);

    propName = "ForbiddenCharacters";
    uno::Reference<i18n::XForbiddenCharacters> xForbiddenCharactersGet;
    uno::Reference<i18n::XForbiddenCharacters> xForbiddenCharactersSet;
    CPPUNIT_ASSERT_MESSAGE("Unable to get PropertyValue: ForbiddenCharacters",
                           xSpreadsheetDocumentSettings->getPropertyValue(propName)
                           >>= xForbiddenCharactersGet);

    lang::Locale aForbiddenCharactersLocale;
    aForbiddenCharactersLocale.Language = "ru";
    aForbiddenCharactersLocale.Country = "RU";

    uno::Reference<i18n::XForbiddenCharacters> xForbiddenCharactersNew;
    aNewValue <<= xForbiddenCharactersNew;
    xSpreadsheetDocumentSettings->setPropertyValue(propName, aNewValue);
    CPPUNIT_ASSERT(xSpreadsheetDocumentSettings->getPropertyValue(propName)
                   >>= xForbiddenCharactersSet);
    CPPUNIT_ASSERT_EQUAL_MESSAGE(
        "Able to set PropertyValue: ForbiddenCharacters",
        xForbiddenCharactersGet->hasForbiddenCharacters(aForbiddenCharactersLocale),
        xForbiddenCharactersSet->hasForbiddenCharacters(aForbiddenCharactersLocale));

    propName = "HasDrawPages";
    testBooleanReadonlyProperty(xSpreadsheetDocumentSettings, propName);

    propName = "CharLocale";
    lang::Locale aCharLocaleGet;
    lang::Locale aCharLocaleSet;
    CPPUNIT_ASSERT_MESSAGE("Unable to get PropertyValue: CharLocale",
                           xSpreadsheetDocumentSettings->getPropertyValue(propName)
                           >>= aCharLocaleGet);

    lang::Locale aCharLocaleNew;
    aCharLocaleNew.Language = "en";
    aCharLocaleNew.Country = "CA";
    aNewValue <<= aCharLocaleNew;
    xSpreadsheetDocumentSettings->setPropertyValue(propName, aNewValue);
    CPPUNIT_ASSERT(xSpreadsheetDocumentSettings->getPropertyValue(propName) >>= aCharLocaleSet);
    CPPUNIT_ASSERT_EQUAL_MESSAGE("Unable to set PropertyValue: CharLocale", aCharLocaleNew.Country,
                                 aCharLocaleSet.Country);

    propName = "CharLocaleAsian";
    lang::Locale aCharLocaleAsianGet;
    lang::Locale aCharLocaleAsianSet;
    CPPUNIT_ASSERT_MESSAGE("Unable to get PropertyValue: CharLocaleAsian",
                           xSpreadsheetDocumentSettings->getPropertyValue(propName)
                           >>= aCharLocaleAsianGet);

    lang::Locale aCharLocaleAsianNew;
    aCharLocaleAsianNew.Language = "en";
    aCharLocaleAsianNew.Country = "CA";
    aNewValue <<= aCharLocaleAsianNew;
    xSpreadsheetDocumentSettings->setPropertyValue(propName, aNewValue);
    CPPUNIT_ASSERT(xSpreadsheetDocumentSettings->getPropertyValue(propName)
                   >>= aCharLocaleAsianSet);
    CPPUNIT_ASSERT_EQUAL_MESSAGE("Unable to set PropertyValue: CharLocaleAsian",
                                 aCharLocaleAsianNew.Country, aCharLocaleAsianSet.Country);

    propName = "CharLocaleComplex";
    lang::Locale aCharLocaleComplexGet;
    lang::Locale aCharLocaleComplexSet;
    CPPUNIT_ASSERT_MESSAGE("Unable to get PropertyValue: CharLocaleComplex",
                           xSpreadsheetDocumentSettings->getPropertyValue(propName)
                           >>= aCharLocaleComplexGet);

    lang::Locale aCharLocaleComplexNew;
    aCharLocaleComplexNew.Language = "en";
    aCharLocaleComplexNew.Country = "CA";
    aNewValue <<= aCharLocaleComplexNew;
    xSpreadsheetDocumentSettings->setPropertyValue(propName, aNewValue);
    CPPUNIT_ASSERT(xSpreadsheetDocumentSettings->getPropertyValue(propName)
                   >>= aCharLocaleComplexSet);
    CPPUNIT_ASSERT_EQUAL_MESSAGE("Unable to set PropertyValue: CharLocaleComplex",
                                 aCharLocaleComplexNew.Country, aCharLocaleComplexSet.Country);

    propName = "IsLoaded";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);

    propName = "IsUndoEnabled";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);

    propName = "IsAdjustHeightEnabled";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);

    propName = "IsExecuteLinkEnabled";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);

    propName = "ReferenceDevice";
    uno::Reference<awt::XDevice> xDeviceGet;
    uno::Reference<awt::XDevice> xDeviceSet;
    CPPUNIT_ASSERT_MESSAGE("Unable to get PropertyValue: ReferenceDevice",
                           xSpreadsheetDocumentSettings->getPropertyValue(propName) >>= xDeviceGet);

    uno::Reference<awt::XDevice> xDeviceNew;
    aNewValue <<= xDeviceNew;
    xSpreadsheetDocumentSettings->setPropertyValue(propName, aNewValue);
    CPPUNIT_ASSERT(xSpreadsheetDocumentSettings->getPropertyValue(propName) >>= xDeviceSet);
    CPPUNIT_ASSERT_EQUAL_MESSAGE("Able to set PropertyValue: ReferenceDevice",
                                 xDeviceSet->getInfo().Width, xDeviceGet->getInfo().Width);

    propName = "RecordChanges";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);

    propName = "IsRecordChangesProtected";
    testBooleanReadonlyProperty(xSpreadsheetDocumentSettings, propName);

    propName = "Wildcards";
    testBooleanProperty(xSpreadsheetDocumentSettings, propName);
}
}

/* vim:set shiftwidth=4 softtabstop=4 expandtab cinoptions=b1,g0,N-s cinkeys+=0=break: */
